pkg/chains/storage/grafeas/grafeas_test.go

```diff
@@ -0,0 +1,415 @@
+	}
+
+	objectIdentifier, err := backend.getResourceURI(test.args.opts)
+	if (err != nil) != test.wantErr {
+		t.Fatalf("Backend.getResourceURI() failed. error:%v, wantErr:%v", err, test.wantErr)
+	}
+
+	// check signature
+	expect_signature := map[string][]string{objectIdentifier: []string{test.args.signature}}
+	got_signature, err := backend.RetrieveSignatures(ctx, test.args.opts)
+	if err != nil {
+		t.Fatal("Backend.RetrieveSignatures() failed. error:", err)
+	}
+
+	if !cmp.Equal(got_signature, expect_signature) && !test.wantErr {
+		t.Errorf("Wrong signature object received, got=%v", cmp.Diff(got_signature, expect_signature))
+	}
+
+	// check payload
+	expect_payload := map[string]string{objectIdentifier: string(test.args.payload)}
+	got_payload, err := backend.RetrievePayloads(ctx, test.args.opts)
+	if err != nil {
+		t.Fatalf("RetrievePayloads.RetrievePayloads() failed. error:%v", err)
+	}
+
+	if !cmp.Equal(got_payload, expect_payload) && !test.wantErr {
+		t.Errorf("Wrong payload object received, got=%s", cmp.Diff(got_payload, expect_payload))
+	}
+}
+
+// test occurrences are generated correctly
+func testListOccurrences(ctx context.Context, t *testing.T, client pb.GrafeasV1Beta1Client) {
+	wantedOccurrences := getExpectedOccurrences()
+
+	// 1. test empty filter - all occurrences
+	// ProjectID field in ListOccurrencesRequest doesn't matter here because we assume there is only one project in the mocked server.
+	got, err := client.ListOccurrences(ctx, &pb.ListOccurrencesRequest{})
+	if err != nil {
+		t.Fatal("Failed to call ListOccurrences. error ", err)
+	}
+	wantedResponse := &pb.ListOccurrencesResponse{Occurrences: wantedOccurrences}
+	if !cmp.Equal(got, wantedResponse, protocmp.Transform()) {
+		t.Errorf("Wrong list of occurrences received for empty filter, got=%s", cmp.Diff(got, wantedResponse, protocmp.Transform()))
+	}
+
+	// 2. test multiple chained filter - return multiple occurrences
+	got, err = client.ListOccurrences(ctx, &pb.ListOccurrencesRequest{
+		Filter: `resourceUrl="/apis/tekton.dev/v1beta1/namespaces/foo1/TaskRun/bar1@uid1" OR resourceUrl="gcr.io/test/kaniko-chains1@sha256:cfe4f0bf41c80609214f9b8ec0408b1afb28b3ced343b944aaa05d47caba3e00"`,
+	})
+	if err != nil {
+		t.Fatal("Failed to call ListOccurrences. error ", err)
+	}
+
+	// wanted response should be same as #1 as there are only two occurrences stored
+	if !cmp.Equal(got, wantedResponse, protocmp.Transform()) {
+		t.Errorf("Wrong list of occurrences received for multiple filters, got=%s", cmp.Diff(got, wantedResponse, protocmp.Transform()))
+	}
+
+	// 3. test a single filter - return one occurrence
+	got, err = client.ListOccurrences(ctx, &pb.ListOccurrencesRequest{
+		Filter: `resourceUrl="gcr.io/test/kaniko-chains1@sha256:cfe4f0bf41c80609214f9b8ec0408b1afb28b3ced343b944aaa05d47caba3e00" OR resourceUrl="/apis/tekton.dev/v1beta1/namespaces/foo2/TaskRun/bar2@uid2"`,
+	})
+	if err != nil {
+		t.Fatal("Failed to call ListOccurrences. error ", err)
+	}
```

Comment:
We're overloading the StorePayload tests a bit too much here - it probably makes more sense to break out the "is the filter working properly" tests into their own TestFunc.

Reply:
Make sense to me!
How about we just leave the test for empty filter here?  And then we make another test func in which we test different filters but still use the two sample occurrences for testing (otherwise feel like there are too much lines for manually creating occurrence objects :D